File: app/src/main/java/online/fatbook/fatbookapp/ui/fragment/RecipeCreateFragment.java
package online.fatbook.fatbookapp.ui.fragment;

import android.Manifest;
import android.app.Activity;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.fragment.NavHostFragment;
import androidx.recyclerview.widget.RecyclerView;

import online.fatbook.fatbookapp.R;
import online.fatbook.fatbookapp.core.Recipe;
import online.fatbook.fatbookapp.core.RecipeIngredient;
import online.fatbook.fatbookapp.databinding.FragmentRecipeCreateBinding;
import online.fatbook.fatbookapp.retrofit.RetrofitFactory;
import online.fatbook.fatbookapp.ui.adapters.ViewRecipeIngredientAdapter;
import online.fatbook.fatbookapp.ui.listeners.OnRecipeViewDeleteIngredient;
import online.fatbook.fatbookapp.ui.viewmodel.RecipeViewModel;
import online.fatbook.fatbookapp.ui.viewmodel.UserViewModel;
import online.fatbook.fatbookapp.util.FileUtils;
import online.fatbook.fatbookapp.util.KeyboardActionUtil;
import online.fatbook.fatbookapp.util.RecipeUtils;

import org.apache.commons.lang3.StringUtils;

import java.io.File;
import java.util.ArrayList;
import java.util.Date;
import java.util.logging.Level;

import lombok.extern.java.Log;
import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.RequestBody;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

@Log
public class RecipeCreateFragment extends Fragment implements OnRecipeViewDeleteIngredient {

    private FragmentRecipeCreateBinding binding;

    private UserViewModel userViewModel;

    private RecipeViewModel recipeViewModel;

    private Recipe recipe;

    private ViewRecipeIngredientAdapter adapter;

    private ActivityResultLauncher<String> choosePhotoFromGallery;

    private File recipePhoto;

    private Uri selectedImageUri;

    private static final int REQUEST_EXTERNAL_STORAGE = 1;
    private static final String[] PERMISSIONS_STORAGE = {Manifest.permission.READ_EXTERNAL_STORAGE, Manifest.permission.WRITE_EXTERNAL_STORAGE};

    @Override
    public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        requireActivity().getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_PAN);

        userViewModel = new ViewModelProvider(requireActivity()).get(UserViewModel.class);
        recipeViewModel = new ViewModelProvider(requireActivity()).get(RecipeViewModel.class);

        recipeViewModel.setSelectedRecipeIngredients(new ArrayList<>());

        try {
            choosePhotoFromGallery = registerForActivityResult(new ActivityResultContracts.GetContent(), uri -> {
                verifyStoragePermissions(requireActivity());
                if (uri != null) {
                    selectedImageUri = uri;
                    String path = FileUtils.getPath(requireContext(), selectedImageUri);
                    recipePhoto = new File(path);
                    binding.imageViewRecipeCreateImage.setImageURI(selectedImageUri);
                }
            });
        } catch (Exception e) {
            e.printStackTrace();
        }

        binding.buttonRecipeCreateImage.setOnClickListener(_view -> {
            verifyStoragePermissions(requireActivity());
            choosePhotoFromGallery.launch("image/*");
        });

        binding.buttonRecipeAddSave.setOnClickListener(_view -> {
            saveRecipe();
        });

        binding.buttonRecipeAddIngredientAdd.setOnClickListener(_view -> {
            NavHostFragment.findNavController(this).navigate(R.id.navigation_add_ingredient);
        });

        setupAdapter();

        binding.editTextRecipeAddTitle.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                niceCheck();
            }

            @Override
            public void afterTextChanged(Editable editable) {
            }
        });

        binding.editTextRecipeAddDescription.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                niceCheck();
            }

            @Override
            public void afterTextChanged(Editable editable) {
            }
        });
    }

    private void saveRecipe() {
        recipe.setName(binding.editTextRecipeAddTitle.getText().toString());
        recipe.setDescription(binding.editTextRecipeAddDescription.getText().toString());
        recipe.setImage(StringUtils.EMPTY);
        recipe.setAuthor(userViewModel.getUser().getValue().getLogin());
        recipe.setCreateDate(RecipeUtils.regDateFormat.format(new Date()));
        save();
        NavHostFragment.findNavController(this).popBackStack();
    }

    private void save() {
        RetrofitFactory.apiServiceClient().recipeCreate(recipe).enqueue(new Callback<Recipe>() {
            @Override
            public void onResponse(@NonNull Call<Recipe> call, @NonNull Response<Recipe> response) {
                if (response.code() == 200) {
                    log.log(Level.INFO, "recipe create SUCCESS");
                    recipe = response.body();
                    uploadImage();
                } else {
                    log.log(Level.INFO, "recipe create FAILED " + response.code());
                }
            }

            @Override
            public void onFailure(@NonNull Call<Recipe> call, @NonNull Throwable t) {
                log.log(Level.INFO, "recipe create FAILED");
            }
        });
    }

    private void uploadImage() {
        if (recipePhoto != null) {
            try {
                RequestBody requestFile = RequestBody.create(MediaType.parse("image/*"), recipePhoto);
                String fileName = "image" + recipePhoto.getName().substring(recipePhoto.getName().indexOf('.'));
                MultipartBody.Part file = MultipartBody.Part.createFormData("file", fileName, requestFile);

                RetrofitFactory.apiServiceClient().uploadImage(file, FileUtils.TAG_RECIPE, recipe.getIdentifier()).enqueue(new Callback<Recipe>() {
                    @Override
                    public void onResponse(@NonNull Call<Recipe> call, @NonNull Response<Recipe> response) {
                        if (response.code() == 200) {
                            log.log(Level.INFO, "image add SUCCESS");
                        } else {
                            log.log(Level.INFO, "image add FAILED " + response.code());
                        }
                    }

                    @Override
                    public void onFailure(@NonNull Call<Recipe> call, @NonNull Throwable t) {
                        log.log(Level.INFO, "image add FAILED");
                    }
                });
            } catch (Exception e) {
                log.log(Level.INFO, e.toString());
                e.printStackTrace();
            }
        }
    }

    private void initRecipe() {
        recipe = new Recipe();
        recipe.setIngredients(new ArrayList<>());
        recipe.setForks(0);
        recipe.setIdentifier(0L);
    }

    private void niceCheck() {
        if (StringUtils.isNotEmpty(binding.editTextRecipeAddTitle.getText().toString())
                && StringUtils.isNotEmpty(binding.editTextRecipeAddDescription.getText().toString())
                && !recipe.getIngredients().isEmpty()) {
            binding.buttonRecipeAddSave.setEnabled(true);
            binding.buttonRecipeAddSave.setBackgroundTintList(ContextCompat.getColorStateList(requireContext(), R.color.color_pink_a200));
        } else {
            binding.buttonRecipeAddSave.setEnabled(false);
            binding.buttonRecipeAddSave.setBackgroundTintList(ContextCompat.getColorStateList(requireContext(), R.color.color_blue_grey_200));
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        binding.getRoot().getViewTreeObserver().addOnGlobalLayoutListener(new KeyboardActionUtil(binding.getRoot(), requireActivity(), binding.viewRecipeCreate).listenerForAdjustPan);
        if (recipeViewModel.getSelectedRecipeIngredient().getValue() != null) {
            recipe.getIngredients().add(recipeViewModel.getSelectedRecipeIngredient().getValue());
            adapter.setData(recipe.getIngredients());
            adapter.notifyDataSetChanged();
            recipeViewModel.setSelectedRecipeIngredient(null);
        }
        if (selectedImageUri != null) {
            binding.imageViewRecipeCreateImage.setImageURI(selectedImageUri);
        }
        niceCheck();
    }

    private void setupAdapter() {
        if (recipe == null) {
            initRecipe();
        }
        RecyclerView recyclerView = binding.rvRecipeAddIngredients;
        adapter = new ViewRecipeIngredientAdapter(binding.getRoot().getContext(), recipe.getIngredients(), this);
        adapter.setEditMode(true);
        recyclerView.setAdapter(adapter);
    }

    @Override
    public void onDeleteIngredientClick(RecipeIngredient recipeIngredient, int position) {
        recipe.getIngredients().remove(recipeIngredient);
        adapter.setData(recipe.getIngredients());
        adapter.notifyItemRemoved(position);
        niceCheck();
    }

    private static void verifyStoragePermissions(Activity activity) {
        int permission = ActivityCompat.checkSelfPermission(activity, Manifest.permission.WRITE_EXTERNAL_STORAGE);

        if (permission != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(activity, PERMISSIONS_STORAGE, REQUEST_EXTERNAL_STORAGE);
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = FragmentRecipeCreateBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onPause() {
        super.onPause();
        binding.getRoot().getViewTreeObserver().removeOnGlobalLayoutListener(new KeyboardActionUtil(binding.getRoot(), requireActivity(), binding.viewRecipeCreate).listenerForAdjustPan);
    }
}

File: app/src/main/java/online/fatbook/fatbookapp/ui/fragment/RecipeCreateFragment.kt
package online.fatbook.fatbookapp.ui.fragment

import android.Manifest
import android.app.Activity
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.*
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts.GetContent
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.fragment.NavHostFragment
import lombok.extern.java.Log
import okhttp3.MediaType
import okhttp3.MultipartBody
import okhttp3.RequestBody
import online.fatbook.fatbookapp.R
import online.fatbook.fatbookapp.core.Recipe
import online.fatbook.fatbookapp.core.RecipeIngredient
import online.fatbook.fatbookapp.databinding.FragmentRecipeCreateBinding
import online.fatbook.fatbookapp.retrofit.RetrofitFactory
import online.fatbook.fatbookapp.ui.adapters.ViewRecipeIngredientAdapter
import online.fatbook.fatbookapp.ui.listeners.OnRecipeViewDeleteIngredient
import online.fatbook.fatbookapp.ui.viewmodel.RecipeViewModel
import online.fatbook.fatbookapp.ui.viewmodel.UserViewModel
import online.fatbook.fatbookapp.util.FileUtils
import online.fatbook.fatbookapp.util.KeyboardActionUtil
import online.fatbook.fatbookapp.util.RecipeUtils
import org.apache.commons.lang3.StringUtils
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.io.File
import java.util.*
import java.util.logging.Level

class RecipeCreateFragment : Fragment(), OnRecipeViewDeleteIngredient {
    private var binding: FragmentRecipeCreateBinding? = null
    private var userViewModel: UserViewModel? = null
    private var recipeViewModel: RecipeViewModel? = null
    private var recipe: Recipe? = null
    private var adapter: ViewRecipeIngredientAdapter? = null
    private var choosePhotoFromGallery: ActivityResultLauncher<String>? = null
    private var recipePhoto: File? = null
    private var selectedImageUri: Uri? = null

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        requireActivity().window.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_PAN)
        userViewModel = ViewModelProvider(requireActivity())[UserViewModel::class.java]
        recipeViewModel = ViewModelProvider(requireActivity())[RecipeViewModel::class.java]
        recipeViewModel!!.selectedRecipeIngredients.value = ArrayList()
        try {
            choosePhotoFromGallery = registerForActivityResult(GetContent()) { uri: Uri? ->
                verifyStoragePermissions(requireActivity())
                if (uri != null) {
                    selectedImageUri = uri
                    val path = FileUtils.getPath(requireContext(), selectedImageUri!!)
                    recipePhoto = File(path)
                    binding!!.imageViewRecipeCreateImage.setImageURI(selectedImageUri)
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        binding!!.buttonRecipeCreateImage.setOnClickListener { _view: View? ->
            verifyStoragePermissions(requireActivity())
            choosePhotoFromGallery!!.launch("image/*")
        }
        binding!!.buttonRecipeAddSave.setOnClickListener { _view: View? -> saveRecipe() }
        binding!!.buttonRecipeAddIngredientAdd.setOnClickListener { _view: View? ->
            NavHostFragment.findNavController(
                this
            ).navigate(R.id.navigation_add_ingredient)
        }
        setupAdapter()
        binding!!.editTextRecipeAddTitle.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                niceCheck()
            }

            override fun afterTextChanged(editable: Editable) {}
        })
        binding!!.editTextRecipeAddDescription.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                niceCheck()
            }

            override fun afterTextChanged(editable: Editable) {}
        })
    }

    private fun saveRecipe() {
        recipe!!.name = binding!!.editTextRecipeAddTitle.text.toString()
        recipe!!.description = binding!!.editTextRecipeAddDescription.text.toString()
        recipe!!.image = StringUtils.EMPTY
        recipe!!.author = userViewModel!!.user.value!!.login
        recipe!!.createDate = RecipeUtils.regDateFormat.format(Date())
        save()
        NavHostFragment.findNavController(this).popBackStack()
    }

    private fun save() {
        RetrofitFactory.apiServiceClient().recipeCreate(recipe).enqueue(object : Callback<Recipe?> {
            override fun onResponse(call: Call<Recipe?>, response: Response<Recipe?>) {
                if (response.code() == 200) {
//                    RecipeCreateFragment.log.log(Level.INFO, "recipe create SUCCESS")
                    recipe = response.body()
                    uploadImage()
                } else {
//                    RecipeCreateFragment.log.log(
//                        Level.INFO,
//                        "recipe create FAILED " + response.code()
//                    )
                }
            }

            override fun onFailure(call: Call<Recipe?>, t: Throwable) {
//                RecipeCreateFragment.log.log(Level.INFO, "recipe create FAILED")
            }
        })
    }

    private fun uploadImage() {
        if (recipePhoto != null) {
            try {
                val requestFile = RequestBody.create(MediaType.parse("image/*"), recipePhoto)
                val fileName = "image" + recipePhoto!!.name.substring(
                    recipePhoto!!.name.indexOf('.')
                )
                val file = MultipartBody.Part.createFormData("file", fileName, requestFile)
                RetrofitFactory.apiServiceClient()
                    .uploadImage(file, FileUtils.TAG_RECIPE, recipe!!.identifier)
                    .enqueue(object : Callback<Recipe?> {
                        override fun onResponse(call: Call<Recipe?>, response: Response<Recipe?>) {
                            if (response.code() == 200) {
//                                RecipeCreateFragment.log.log(Level.INFO, "image add SUCCESS")
                            } else {
//                                RecipeCreateFragment.log.log(
//                                    Level.INFO,
//                                    "image add FAILED " + response.code()
//                                )
                            }
                        }

                        override fun onFailure(call: Call<Recipe?>, t: Throwable) {
//                            RecipeCreateFragment.log.log(Level.INFO, "image add FAILED")
                        }
                    })
            } catch (e: Exception) {
//                RecipeCreateFragment.log.log(Level.INFO, e.toString())
                e.printStackTrace()
            }
        }
    }

    private fun initRecipe() {
        recipe = Recipe()
        recipe!!.ingredients = ArrayList()
        recipe!!.forks = 0
        recipe!!.identifier = 0L
    }

    private fun niceCheck() {
        if (StringUtils.isNotEmpty(binding!!.editTextRecipeAddTitle.text.toString())
            && StringUtils.isNotEmpty(binding!!.editTextRecipeAddDescription.text.toString())
            && recipe!!.ingredients!!.isNotEmpty()
        ) {
            binding!!.buttonRecipeAddSave.isEnabled = true
            binding!!.buttonRecipeAddSave.backgroundTintList =
                ContextCompat.getColorStateList(requireContext(), R.color.color_pink_a200)
        } else {
            binding!!.buttonRecipeAddSave.isEnabled = false
            binding!!.buttonRecipeAddSave.backgroundTintList =
                ContextCompat.getColorStateList(requireContext(), R.color.color_blue_grey_200)
        }
    }

    override fun onResume() {
        super.onResume()
        binding!!.root.viewTreeObserver.addOnGlobalLayoutListener(
            KeyboardActionUtil(
                binding!!.root, requireActivity(), binding!!.viewRecipeCreate
            ).listenerForAdjustPan
        )
        if (recipeViewModel!!.selectedRecipeIngredient.value != null) {
            recipe!!.ingredients!!.add(recipeViewModel!!.selectedRecipeIngredient.value!!)
            adapter!!.setData(recipe!!.ingredients)
            adapter!!.notifyDataSetChanged()
            recipeViewModel!!.selectedRecipeIngredient.value = null
        }
        if (selectedImageUri != null) {
            binding!!.imageViewRecipeCreateImage.setImageURI(selectedImageUri)
        }
        niceCheck()
    }

    private fun setupAdapter() {
        if (recipe == null) {
            initRecipe()
        }
        val recyclerView = binding!!.rvRecipeAddIngredients
        adapter!!.setData(recipe!!.ingredients)
        adapter!!.setClickListener(this)
        adapter!!.setEditMode(true)
        recyclerView.adapter = adapter
    }

    override fun onDeleteIngredientClick(recipeIngredient: RecipeIngredient?, position: Int) {
        recipe!!.ingredients!!.remove(recipeIngredient)
        adapter!!.setData(recipe!!.ingredients)
        adapter!!.notifyItemRemoved(position)
        niceCheck()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        binding = null
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentRecipeCreateBinding.inflate(inflater, container, false)
        return binding!!.root
    }

    override fun onPause() {
        super.onPause()
        binding!!.root.viewTreeObserver.removeOnGlobalLayoutListener(
            KeyboardActionUtil(
                binding!!.root, requireActivity(), binding!!.viewRecipeCreate
            ).listenerForAdjustPan
        )
    }

    companion object {
        private const val REQUEST_EXTERNAL_STORAGE = 1
        private val PERMISSIONS_STORAGE = arrayOf(
            Manifest.permission.READ_EXTERNAL_STORAGE,
            Manifest.permission.WRITE_EXTERNAL_STORAGE
        )

        private fun verifyStoragePermissions(activity: Activity) {
            val permission = ActivityCompat.checkSelfPermission(
                activity,
                Manifest.permission.WRITE_EXTERNAL_STORAGE
            )
            if (permission != PackageManager.PERMISSION_GRANTED) {
                ActivityCompat.requestPermissions(
                    activity,
                    PERMISSIONS_STORAGE,
                    REQUEST_EXTERNAL_STORAGE
                )
            }
        }
    }
}